addons/survey/controllers/main.py

```diff
@@ -25,21 +26,35 @@ class Survey(http.Controller):
     # ACCESS
     # ------------------------------------------------------------
 
-    def _fetch_from_access_token(self, survey_token, answer_token):
+    def fetch_answer(self, survey_token, answer_token, survey_id=False, answer_id=False):
         """ Check that given token matches an answer from the given survey_id.
         Returns a sudo-ed browse record of survey in order to avoid access rights
         issues now that access is granted through token. """
-        survey_sudo = request.env['survey.survey'].with_context(active_test=False).sudo().search([('access_token', '=', survey_token)])
-        if not answer_token:
+        if survey_id:
+            survey_sudo = request.env['survey.survey'].with_context(active_test=False).sudo().browse(survey_id).exists()
+        else:
+            survey_sudo = request.env['survey.survey'].with_context(active_test=False).sudo().search([('access_token', '=', survey_token)])
+
+        if survey_sudo.exists() and not consteq(survey_sudo.access_token, survey_token):
+            raise NotFound()
+
+        if not answer_token and not answer_id:
             answer_sudo = request.env['survey.user_input'].sudo()
+
+        if answer_id:
+            answer_sudo = request.env['survey.user_input'].with_context().sudo().browse(answer_id).exists()
         else:
```

Comment:
The first conditional block never matters as the value is overwritten in either branch of the following conditional block.

But you don't want to do the search as `answer_token = False` could match any random record. So I would return early in the first block.